File: Dynamic Programming/42.trapping_rain_water.js
// 做完238回来看的题目，discuss里说这两题有思维上的联系
// 多复习，多找题目之间的联系
// dp解法其实就是two-way pass，produce maxHeight to left/right, 对于每一个cell, 其储水量为min(maxRight, maxLeft) - height[cell undex]
// 这是trapping water的其中一种解法，最好理解的一种解法
// Time Complexity: O(n)
// Space Complexity: O(n)

/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    if (height === null || height.length === 0) return 0;

    let left = new Array(height.length);
    let right = new Array(height.length);

    left[0] = 0;
    for (let i = 1; i < left.length; i++) {
        left[i] = Math.max(left[i - 1], height[i - 1]);
    }

    right[height.length - 1] = 0;

    for (let i = height.length - 2; i >= 0; i--) {
        right[i] = Math.max(right[i + 1], height[i + 1]);
    }

    
    let res = 0;
    for (let i = 0; i < height.length; i++) {
        if (Math.min(left[i], right[i]) > height[i]) {
            res += Math.min(left[i], right[i]) - height[i]   
        }
    }

    return res;
};

File: Java/42.trapping_rain_water.java
// Date: 11-18-2019
// Tag: Two Pointers

class Solution {
    public int trap(int[] height) {
        if (height == null || height.length == 0) return 0;
        int res = 0;
        int maxLeft = Integer.MIN_VALUE;
        int maxRight = Integer.MIN_VALUE;

        int left = 0;
        int right = height.length - 1;

        while (left < right) {
            int leftHeight = height[left];
            int rightHeight = height[right];

            if (leftHeight < rightHeight) {
                maxLeft = Math.max(leftHeight, maxLeft);
                res += (maxLeft - leftHeight);
                left++;
            } else {
                maxRight = Math.max(rightHeight, maxRight);
                res += (maxRight - rightHeight);
                right--;
            }
        }

        return res;
    }
}

// Dynamic Programming Solution
// Time Complexity: O(3n) => O(n)
// Space Complexity: O(n)
class Solution {
    public int trap(int[] height) {
        if (height == null || height.length == 0) return 0;
        int res = 0;

        int[] leftMax = new int[height.length];
        leftMax[0] = 0;
        for (int i = 1; i < height.length; i++) {
            leftMax[i] = Math.max(height[i - 1], leftMax[i - 1]);
        }

        int[] rightMax = new int[height.length];
        rightMax[rightMax.length - 1] = 0;

        for (int i = height.length - 2; i >=0; i--) {
            rightMax[i] = Math.max(height[i + 1], rightMax[i + 1]);
        }

        for (int i = 0; i < height.length; i++) {
            int min = Math.min(leftMax[i], rightMax[i]);

            if (min <= height[i]) continue;
            else res += height[i] - min;
        }

        return res;
    }
}

// Stack Solution
// 这个SOLUTION不是那么熟练
// Time COMPLEXITY: O(N)
// Space COMPLEXITY: O(N)
class Solution {
    public int trap(int[] height) {
        if (height == null || height.length == 0) return 0;
        int res = 0;

        Stack<Integer> stack = new Stack<>();

        int current = 0;
        while (current < height.length) {
            while (!stack.isEmpty() && height[current] > height[stack.peek()]) {
                int top = stack.pop();
                if (stack.isEmpty()) break;

                int distance = current - stack.peek() - 1;
                int bound = Math.min(height[current], height[stack.peek()]) - height[top];
                res += bound * distance;
            }
            stack.push(current++);
        }

        return res;
    }
}